### ISSUE TITLE
5 issues with `CompositeAlgorithm`s : bad `unwrap_alg` calls, `alg_cache` with more than 2 algs, `__init` and timestepping issues with mixed-adaptivity `Composite`s

### ISSUE THREAD

**Issue by brainsMAKER** [SciML/OrdinaryDiffEq.jl#1993]
## Issue description

Found 5 different issues when working with `CompositeAlgorithm`s:

1. When switching algorithm with `choose_algorithm!`, `initialize!(integrator,cache)` gets called before modifying `integrator.cache.current`.
    This results in errors when `initialize!` makes use of `alg=unwrap_alg(integrator,is_stiff)` because then `alg` may not correspond to `cache` (e.g. switching from `Tsit5()` to `VernN()` ...).

2. `cache.caches` for `CompositeAlgorithm`s with more than 2 algorithms don't get generated properly (only the first 2 algorithms) because of bad function signatures for `alg_cache`.
    This usually doesn't result in errors when solving thanks to a bad `@inbounds`, but it can.
    It mostly means that 3rd and above algorithms of the `Composite` never get used.
    It does lead to errors if `alg=unwrap_alg` gets used (`alg` may not correspond to `cache`) in `initialize!` or `perform_step!`.

3. When using `CompositeAlgorithm`s with mixed adaptivity, `DiffEqBase.__init` errors out because `qoldinit==0` by default leading to `InexactError : Int64(x::Rational{Int64})` in the call to `default_controller`.

4.  When using `CompositeAlgorithm`s with mixed adaptivity, other default behaviors and argument values in `DiffEqBase.__init` (`dtmin`, `gamma_default`, checks...) lead to issues like `dt`->0  with adaptive timestepping (which does occur because `integrator.opts.adaptive` returns to default value after multiple switches).

5.  When using `CompositeAlgorithm`s the default controller parameters correspond to the first algorithm in `alg.algs`. When using mixed adaptivity `Composite`s, this can easily lead to either `dt`->0 while using an algorithm designed either for adaptive timestepping or not, or `dt=cste` while using a adaptive algorithm (unexpected behavior I believe).

Related issues/PRs : #317, #322, #1265, #1618, #1625, #1847, #1892

## Working example

This is a non-minimal working example :

```julia
using OrdinaryDiffEq, Test

##################################################
# Definitions

# issues with CompositeAlgorithm
brk3 = true             # OrdinaryDiffEq.choose_algorithm! → unwrap_alg in initialize! issue
brk4 = true             # OrdinaryDiffEq.alg_cache issue (with more than 2 algs)

# issues with mixed adaptivity CompositeAlgorithm
brk5 = true             # default parameters type issues in OrdinaryDiffEq.init!
brk6 = (true, true)     # adaptive timestepping issues

if true                 # to prevent unnecesary compilation

    # aliases
    psm(x...) = printstyled(x...; color=:magenta)
    OS = ODESolution

    # problem definitions
    tspan = (0.0, 10.0)

    ftest1D(u,p,t) = u+20cos(5t)
    ftest3D!(du,u,p,t) = @. (du = (-1,0.5,1)*u + 20*cos(5t))

    prb_ODE_1D = ODEProblem(ftest1D, 1.0, tspan)
    prb_ODE_3D = ODEProblem(ftest3D!, ones(3), tspan)

    _choice_function(t::Real) = Int(cos(t)<0)+1
    choice_function(integrator) = _choice_function(integrator.t)            # ==1 or ==2
    choice_function_p1(integrator) = _choice_function(integrator.t)+1       # ==2 or ==3

end

##################################################
# Testing ODE solving with CompositeAlgorithm

@testset verbose=true "Solving ODE with CompositeAlgorithm ($str)" for (str, prb) in
        zip(["1D", "3D"], [prb_ODE_1D, prb_ODE_3D])

    printstyled("\nTests for CompositeAlgorithm ($str):\n"; color=:red, bold=true)

    # 2 adaptive algs
    @testset "  with 2 adaptive algs" begin
        # This one passes of course
        psm("CompositeAlgorithm((Tsit5(),Rodas4()),...)\n")
            alg_switch1 = CompositeAlgorithm((Tsit5(),Rodas4()), choice_function)
            @test @time(solve(prb, alg_switch1)) isa OS
            sol_ODE__ = solve(prb, alg_switch1)
            @test _choice_function.(sol_ODE__.t[1:end-1]) == sol_ODE__.alg_choice[2:end]

        # Error with Vern
        psm("CompositeAlgorithm((Tsit5(),Vern7()),...)\n")
            alg_switch2 = CompositeAlgorithm((Tsit5(),Vern7()), choice_function)
            @test @time(solve(prb, alg_switch2)) isa OS                     broken=brk3
            # problem from unwrap_alg called inside initialize! inside choose_algorithm

        # Error with BS5
        psm("CompositeAlgorithm((Tsit5(),BS5()),...)\n")
            alg_switch3 = CompositeAlgorithm((Tsit5(),BS5()), choice_function)
            @test @time(solve(prb, alg_switch3)) isa OS                     broken=brk3
            # unwrap_alg before updating cache.current issue

        # Error with AitkenNeville
        psm("CompositeAlgorithm((Tsit5(),AitkenNeville()),...)\n")
            alg_switch4 = CompositeAlgorithm((Tsit5(),AitkenNeville()), choice_function)
            @test @time(solve(prb, alg_switch4)) isa OS                     broken=brk3
            # unwrap_alg before updating cache.current issue
    end

    # Errors with 3 algs (including Vern)
    psm("CompositeAlgorithm((Tsit5(),Vern7(),Rodas4()),...) in ≠ orders\n")
    @testset "  with 3 algs (including Vern)" begin
        alg_switch5a = CompositeAlgorithm((Tsit5(),Vern7(),Rodas4()), choice_function)
        @test @time(solve(prb, alg_switch5a)) isa OS                         broken=brk3
        # unwrap_alg before updating cache.current issue
        alg_switch5b = CompositeAlgorithm((Tsit5(),Rodas4(),Vern7()), choice_function)
        @test @time(solve(prb, alg_switch5b)) isa OS
        # passes because does not switch to Vern (expected here)
        inte_3_algs = init(prb, alg_switch5b)
        @test length(inte_3_algs.cache.caches)==3                           broken=brk4
        # not generating 3 caches !
        alg_switch5c = CompositeAlgorithm((Vern7(),Tsit5(),Rodas4()), choice_function_p1)
        @test @time(solve(prb, alg_switch5c)) isa OS
        # passes because never actually switches to Rodas4 because of bad @inbounds
        alg_switch5d = CompositeAlgorithm((Tsit5(),Vern7(),Rodas4()), choice_function_p1)
        @test @time(solve(prb, alg_switch5d)) isa OS                        broken=brk4
        # unwrap_alg with cache.current==3 passes but then does not match Vern7Cache

    end

    # Passing with two fixed time-step methods
    psm("CompositeAlgorithm((ImplicitMidpoint(),ABM54()),...)\n")
    @testset "  with two fixed time-step algs" begin
        alg_switch6 = CompositeAlgorithm((ImplicitMidpoint(),ABM54()), choice_function)
        @test @time(solve(prb, alg_switch6; dt=0.1)) isa OS
        # did not find one with unwrap_alg issues
    end

    # Errors with mixed adaptivity
    @testset verbose=true "  with adaptive and not" begin

        psm("CompositeAlgorithm((Tsit5(),ABM54()),...)\n")
        @testset "    adaptive first" begin
            alg_switch7 = CompositeAlgorithm((Tsit5(),ABM54()), choice_function)
            @test @time(solve(prb, alg_switch7; dt=0.1)) isa OS             broken=brk5
            # problem in __init -> qoldinit = 0 -> then Int64(2//25) in default_controller
            @time(sol3 = solve(prb, alg_switch7; dt=0.1, qoldinit=1//10^4))
            @test sol3.retcode == ReturnCode.Success                        broken=brk6[1]
            # different problems -> with adaptive timestepping
        end

        psm("CompositeAlgorithm((ABM54(),Tsit5()),...)\n")
        @testset "    fixed-step first" begin
            alg_switch8 = CompositeAlgorithm((ABM54(),Tsit5()), choice_function)
            @test @time(solve(prb, alg_switch8; dt=0.1)) isa OS             broken=brk5
            # problem in __init
            @time(sol4 = solve(prb, alg_switch8; dt=0.1, qoldinit=1//10^4, adaptive=true))
            @test sol4.retcode == ReturnCode.Success                        broken=brk6[2]
            # different problem -> with adaptive timestepping
        end

    end
end
```
**Here is the screenshot of the broken result in 6.53.3 :**

![issue3_broken](https://github.com/SciML/OrdinaryDiffEq.jl/assets/48725375/cdc456d2-7180-4b5d-85e4-434e61d84981)

## Proposed fixes

Summary of the proposed fixes :

1. If necessary, store the previous value of `cache.current` in a variable for later use and change cache.current at the start of the `if` block in `choose_algorithm!`.

2. Two versions of `alg_cache`, one for 2 algs and one for multiple algs, were already written.  The signatures just need to be exchanged.

- **Mixed adaptivity** I feel like anyone using such `Composite` probably wants to make use of the 2 types of timestepping, and the current behaviour is that they might both occur regardless of the user default. Therefore I propose changing the default values and corresponding checks that lead to errors to match this assumption.
***This behaviour should be reflected in the documentation.***

5. This is partially fixed by the previous fix. Working with the same assumption, I porpose calling `reset_alg_dependent_opts(controller,...)` during integrator initialization and after, in case of mixed adaptivity, ensuring that `integrator.opts.adaptive` does corresponds to the first algorithm used.

**Here is the screenshot of the fixed result (incoming PR) :**
![issue3_fixed](https://github.com/SciML/OrdinaryDiffEq.jl/assets/48725375/9a7108de-7fcd-4482-8413-4a215426caea)

**Comment by ChrisRackauckas** [comment]
> Mixed adaptivity I feel like anyone using such Composite probably wants to make use of the 2 types of timestepping, and the current behaviour is that they might both occur regardless of the user default. Therefore I propose changing the default values and corresponding checks that lead to errors to match this assumption.

The most common usage of CompositeAlgorithm is for the stiffness detection methods which use one non-stiff and one stiff solver, both adaptive, and switches between the two based on the presence of stiffness. 
So >99% of the uses of CompositeAlgorithm both have `adaptive=true`. We should definitely make sure it's robust to more behaviors, but I think it should be clear that the reason why this is a less tested is because it's probably the first case I've seen of someone mixing a method with adaptivity and one without. I'd actually be curious to hear more about how you're using it and how that led to a mixed adaptivity case.

Do let us know if there's any other things that need to be handled with the mixed adaptivity case. The tests around it look pretty solid now (thanks to you).

**Comment by brainsMAKER** [comment]
> The most common usage of CompositeAlgorithm is for the stiffness detection methods which use one non-stiff and one stiff solver, both adaptive, and switches between the two based on the presence of stiffness. So >99% of the uses of CompositeAlgorithm both have `adaptive=true`.  We should definitely make sure it's robust to more behaviors, but I think it should be clear that the reason why this is a less tested is because it's probably the first case I've seen of someone mixing a method with adaptivity and one without. 

This was pretty clear for me, it makes sense this wasn't tested earlier. This is also why I assumed in the changes that anyone using a `Composite` with mixed adaptivity really wanted both the adaptive and non-adaptive to work accordingly.

> I'd actually be curious to hear more about how you're using it and how that led to a mixed adaptivity case.

The truth is that I stumbled on all these issues mostly by chance while trying to be reasonably exhaustive in unit testing.
- I wanted to see if I had indeed improved allocations during interpolation with `CompositeAlg`s 
- I tried `CompositeAlgorithm((Tsit5(),Vern7()),...)` with a periodic choice function because I often use these algorithms in practice. Thus I found the `unwrap_alg` issue.
- Then I tried `CompositeAlgorithm((Tsit5(),Rodas4(),Vern7()),...)` with a periodic 3-choice function and couldn't understand why it did not error, this is how I found the `alg_cache` thing.
- Having fixed these, I figured I now had to try a `Composite` with non-adaptive algs and **THIS** is why I tried mixed adaptivity.

Well, I guess this just highlights the importance (and the pain) of unit-testing.

> I'd actually be curious to hear more about how you're using it and how that led to a mixed adaptivity case.

I have actually thought of reasons to use mixed-adaptivity : some systems have predictable moments of stiffness or sensitivity,  e.g. spiking neurons, time-delay systems, systems with time dependent input,...
These might benefit from the high efficiency of some fixed-step algorithms (with a rather large timestep) during portions of the integration then switching to adaptive or stiff-aware algorithms at strategic times.
(I am thinking in particular of stochastic systems for which certain fixed-step algorithms have pretty nice convergence properties and speed).

> Do let us know if there's any other things that need to be handled with the mixed adaptivity case. The tests around it look pretty solid now (thanks to you).

Well I haven't tested mixed-adaptivity `Composite`s with StochasticDiffEq.jl yet and I believe I might use it soon so I might get back for this.
I certainly hope I fixed as much as possible around this here.
If anyone finds some other issues related to this, don't hesitate to @ me, to see if I can be of some help.
